Applying Numba on plot_kde[WIP]

Ban-zee opened the pull request:
This is the first pull request of my project(applying Numba across ArviZ). Open for review.

ahartikainen commented on arviz/plots/kdeplot.py:
```diff
@@ -4,9 +4,9 @@
 import matplotlib.pyplot as plt
 from scipy.signal import gaussian, convolve, convolve2d  # pylint: disable=no-name-in-module
 from scipy.sparse import coo_matrix
+from numba import njit
```
do we still keep numba as an optional dependency?

aloctavodia replied:
I think we want to keep this optional. not sure how easy is to install numba if not using conda

ahartikainen replied:
I think it is not too bad, but arch and other smaller distros could have a problem

canyon289 commented:
@Ban-zee From my understanding you're still working on this?

Ban-zee commented:
@canyon289 Yes, waiting for feedback.

Ban-zee commented:
Not optional if everything works well

aloctavodia commented on arviz/plots/kdeplot.py:
```diff
@@ -262,6 +262,7 @@ def plot_kde(
     return ax
 
 
+@njit(cache=True)
```
Is this really faster? Last time I tried this (and if my memory is not failing me) this was slower/or did not give any speed-up. That's why numba was restricted to  `_histogram`. It will be nice if every PR related to numba comes with  the speed-up. Maybe we should also have test that ensure the _numbified_ function is faster than the non-modified one.

ahartikainen replied:
That sounds interesting.

Ban-zee replied:
I agree, the improvement isn't much and sometimes the function is a bit slower. This is what I wanted to get checked out at first so as to ensure that the lack of speed-up was not because of some error in the code. Regarding tests, I agree that it is an interesting idea.

Ban-zee commented:
Removed jit from `fast_kde` and `fast_2d_kde` as the enhancement was not much on these functions. Instead of a test as suggested, I added a function which gives the enhancement provided after applying the numba decorator. Up for review.

aloctavodia commented on arviz/utils.py:
```diff
@@ -80,3 +81,19 @@ def wrapper(function):
+    return wrapped
+
+
+def to_numba_or_not_to_numba(function, *args, **kwargs):
```
I understand the idea of this function if for us to manually check if a adding numba is helping or not. Right? I still think we should also have tests in order to detect regressions (not only from our side, but just in case numba changes in some unexpected way). @ColCarroll what do you think about this?

ahartikainen replied:
How accurate timing is on travis. 

Can we send those timings to some server so we can plot against each time they are tested?

Ban-zee replied:
Yes, this gives us an estimate of the numba enhancement. I am thinking of adding a loop in this method so as to get a proper estimate of the time. If everyone is fine with this particular method of testing numba then I'll add the test.

ColCarroll replied:
We use [airspeed velocity](https://github.com/airspeed-velocity/asv) on pymc3. The benchmarks are here: http://pandas.pydata.org/speed/pymc3/, and we check them manually whenever we want to. 

One of the anaconda devs (and all around good person) helped us set that up - I could check whether we can also run `arviz` benchmarks on that machine.

Ban-zee commented:
Added the test and improved the enhancement function. On plot_kde, only histogram showed a considerable enhancement after applying conditional_jit. Jit will have to be disabled for the test to pass or else it would throw an error. If everyone is fine with these changes, then I'll start to apply numba on stats tomorrow as mentioned in my timeline.

canyon289 commented:
I vote optional on Numba, it's hard to install on some systems

Ban-zee commented:
The python tests pass every time on system. My guess is that jit is disabled which leads to this failure. I might be completely wrong as well. :(

aloctavodia commented:
@canyon289 maybe you can help here with the tests. I am not sure what to do.

Ban-zee commented:
NUMBA_DISABLE_JIT is set to 1 in travis.yml, changing it to 0 should resolve the issue.

canyon289 commented:
> @canyon289 maybe you can help here with the tests. I am not sure what to do.

Just saw this sorry. We want to test both circumstances with JIT and without JIT so we'll need another test pipeline.

Ban-zee commented:
Closing this. Opened a new PR for the same.
